internal/jimm/upgrade/upgrade.go

```diff
@@ -0,0 +1,76 @@
+// Copyright 2025 Canonical.
+
+// upgrade package provides functionality to manage the upgrade process
+// for controllers in JIMM.
+package upgrade
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/uuid"
+	"github.com/juju/zaputil/zapctx"
+	"go.uber.org/zap"
+
+	"github.com/canonical/jimm/v3/internal/errors"
+	"github.com/canonical/jimm/v3/internal/jimm/bootstrap"
+	"github.com/canonical/jimm/v3/internal/openfga"
+)
+
+// BootstrapManager defines the bootstrap manager methods required by the upgrade manager.
+type BootstrapManager interface {
+	WaitForJobCompletion(ctx context.Context, user *openfga.User, jobId uuid.UUID) error
+	StartBootstrapJob(ctx context.Context, user *openfga.User, params bootstrap.BootstrapParams) (string, error)
+}
+
+// upgradeManager provides a means to manage controller upgrades within JIMM.
+type upgradeManager struct {
+	bootstrapManager BootstrapManager
+}
+
+// NewUpgradeManager creates a new UpgradeManager instance.
+func NewUpgradeManager(
+	bootstrapManager BootstrapManager,
+) (*upgradeManager, error) {
+	if bootstrapManager == nil {
+		return nil, errors.E("bootstrap manager cannot be nil")
+	}
+	return &upgradeManager{
+		bootstrapManager: bootstrapManager,
+	}, nil
+}
+
+// UpgradeTo upgrades a controller by fetching its configuration and initiating
+// a bootstrap job with that configuration, then waits for the bootstrap to complete.
+func (u *upgradeManager) UpgradeTo(ctx context.Context, user *openfga.User, params UpgradeParams) error {
```

Comment:
Purrrrfecto, I'll plug this https://github.com/canonical/jimm/pull/1740 in once this merges here.